### ISSUE TITLE
[Bug]: changing GUI language changes also decimal separator

### ISSUE THREAD

**Issue by liosc** [apache/hop#2814]
### Apache Hop version?

2.3.0

### Java version?

jdk-11.0.15.10

### Operating system

Windows

### What happened?

Changing the GUI language changes also the decimal separator. I would like to use English language but I need to use Italian decimal separator. If I choose Italian language a number converted to text appears like 10000,23. If I change the language to English, the same number becomes 1000023.0. I have a very simple pipeline to test this but i didn't find a way to attach it here.
I think GUI language should be independent from international settings.

### Issue Priority

Priority: 2

### Issue Component

Component: Hop Gui

**Comment by liosc** [comment]
The biggest problem we have with this issue is related to hash generation. Changing decimal and grouping separator produces a different hash so we are forced to use always the same language.
By the way, it doesn't seem logical to me that when you change the language the decimal and grouping separator change too. As I added in the description, I think GUI language should be independent from international settings.

**Comment by sramazzina** [comment]
@mattcasters I agree with @liosc. Same as in every operating system were these two aspects are managed separately

**Comment by mattcasters** [comment]
This is not the case. You can set the application itself in  locale but in general the data engineers needs to be able to handle numbers in various formats and it's not wise to be cavalier about it.

**Comment by sramazzina** [comment]
Hi @mattcasters . I don’t agree with your point of view. I agree with you that many transforms allow you to change the grouping and decimal separator settings, but this feature addresses a use case that is different from the one @liosc  and I are discussing. The use case covered by the current transforms relates to the fact that a specific file I’m reading may have a different locale than the one set in Hop (implicitly set by the language selection). In this case, however, what we want is to separate the language (English) from the system locale (Italian) because both we and many customers are accustomed to using English as the primary interface language—NOT Italian—for our own convenience. Furthermore, the locale and the interface language are two different things, so it seems correct to me that they can be treated separately. I’m therefore taking the liberty of reopening the issue.

**Comment by mattcasters** [comment]
You can re-open it as many times as you like.  However, then please also take the time to specify what exactly it is that you would to see changed.  I'm not seeing it.

**Comment by hansva** [comment]
.take-issue

**Comment by sramazzina** [comment]
@hansva I've almost finished my implementation let me know what I need to do

**Comment by sramazzina** [comment]
**The problem, put precisely.** Right now the interface language and the regional settings travel together: picking a language also decides your decimal separator, grouping separator, currency symbol and date formats. So there's no way to run an English interface with Italian number formats, which is what a good few of us need day to day.

There's a second half to this that's easy to miss. The language is applied to formats only when the Hop GUI starts up — `hop-run` and `hop-server` never apply it and quietly use the machine's own settings instead. So the very same pipeline can produce different output in the GUI and in production whenever the machine's regional settings don't match the Hop language. If you're hashing or comparing stringified numbers downstream, that's not a cosmetic wrinkle, it's a reproducibility problem. We'd like to fix that too, and carefully.

**What we'd like to build.** A new **Regional settings** tab in the configuration perspective. The existing *Preferred language* dropdown simply moves there, unchanged, and right next to it the regional settings become their own choice with three states:

- **Neither option ticked (the default)** — regional settings come from the selected language, exactly as they do today.
- **Use operating system regional settings** — Hop just inherits the formats from the machine it's running on, whatever the interface language happens to be.
- **Override regional settings** — you pick a locale explicitly, from a list of everything the platform offers rather than only the languages Hop happens to be translated into.

Whenever you make one of those two explicit choices, it's honoured everywhere — the GUI, `hop-run` and `hop-server` alike. Headless runs read the configuration of the machine they're running on, which we think is what people expect in the first place.

**A safety valve for the default case.** Closing the GUI/production gap for people who *haven't* configured anything is trickier than it looks, so we're not doing it silently. The stored language is filled in automatically on first start from the machine's locale — but it falls back to English whenever that language isn't one of the fourteen Hop is translated into, or is still marked beta. On a Swedish or Greek machine, then, the stored language is English while the machine isn't, and simply honouring the configuration would change `hop-run` output for those users without them touching a thing.

So there's a fourth option, **Legacy support for hop-run and hop-server**, enabled by default. While it's on, headless runs keep using the machine's regional settings for the default case, exactly as they do today — upgrade, change nothing, and nothing changes. Turn it off and headless finally agrees with the GUI. It has no effect on the two explicit choices above, which are always honoured either way. It's deliberately transitional: the idea is to drop it in a future major release and make "honour the configuration" simply the behaviour.

The tab also shows a live preview of sample dates, times, numbers and currency for whichever option is selected — much like the equivalent panels in Windows and Linux — so you can see what you're getting before you save it.

**On the per-transform separator options** — @mattcasters, you're right about these, and we're not proposing to touch them. They're the correct tool for their job, and they solve a genuinely different problem: a particular file or field showing up in a format that differs from the environment's. What's missing is the environment-level default underneath, which today can't be set independently of the interface language. The two work together nicely — wherever a per-transform override is set, it still wins.

**Two things we'd rather flag ourselves than have someone discover later.** This kind of setting is global to the running process, so in Hop Web it belongs to the installation rather than to each connected user. That's already true of the interface language today, so it isn't a regression, but it does deserve a line in the docs. And we've deliberately left string collation — and therefore sort order — out of scope. It's arguably a regional setting too, but it changes data-affecting behaviour and deserves a conversation of its own, so we'd rather bring it back as a separate issue once this one has landed.

Backward compatibility is the hard constraint running through all of it: change nothing, and nothing changes. We're covering that with an explicit regression test rather than trusting inspection.

**Comment by mattcasters** [comment]
> picking a language also decides your decimal separator, grouping separator, currency symbol and date formats.

That is simply not true like I stated before.

> So there's no way to run an English interface with Italian number formats, which is what a good few of us need day to day.

What do you even mean by stating something like this?  We all live in Europe Sergio. Italy isn't special you know.

> What we'd like to build. A new Regional settings tab in the configuration perspective. The existing Preferred language dropdown simply moves there, unchanged, and right next to it the regional settings become their own choice with three states:

And then what?  You convinced yourself that you wrote a functioning pipeline or even have some passing tests. And then you'd deploy on a server somewhere, or even with hop-run that doesn't have the option and things would go wrong.

If you're going to tackle this, set aside the AI for a second and do it properly please.  Make it a configuration option for a project lifecycle environment for example.

**Comment by sramazzina** [comment]
Here it is a preview of the new _Regional settings_ dialog

<img width="1613" height="598" alt="Image" src="https://github.com/user-attachments/assets/324acadb-7ec2-4e67-926f-7ba496aec140" />

Tomorrow we will start the first tests to verify the behavior with this new functionality implemented.I will keep you updated

**Comment by hansva** [comment]
Hi all,
Attached is a small project to reproduce some of the issues:
[locality-issues.zip](https://github.com/user-attachments/files/31546953/locality-issues.zip)

Using the UI language of English, some of the workflows will fail if you switch to Dutch, it is flipped the other way.

Most of this can be solved by correctly specifying the group and decimal symbol inside the reading transforms.
But these fields are not mandatory, or not filled in by default, so a lot of people will just go off the defaults that do get switched when changing languages. 

So you could end up in a situation where everything "works on my machine" with `nl_BE` but then when you try to run the same workflow on a Hop Server running under the `en_US` locale, things start failing.

I think a big part of the way forward is also in making these settings very visible, having hop-gui/hop-run/hop-server show what locale settings they are using.

The ROOT fix would be to make the decimal/grouping symbols mandatory and not just use system defaults, but that would be a breaking change, and we would still need some way to define the defaults (US style vs EU style vs whatever other local styles are out there).

As I picked up the issue yesterday, I started an initial analysis/plan of attack on this issue in combination with #5926, as it is a related issue. Issue #5598 is probably also related, but I did not include it in the initial investigation. The investigation can be found [here](https://claude.ai/code/artifact/ee2f8222-e512-4119-9867-73dd00ad370b). DISCLAIMER: It has been done in tandem with Claude

**Comment by mattcasters** [comment]
I haven't been communicating properly in this issue so my appologies to you all for that.  But indeed, "works on my machine" was one of the main concerns plus a series of unintended consequences where a data engineers spends a lot of time chasing string conversion issues, usually right when they want to put something in production on a remote server.  My preference is always to go to the bottom of what we're trying to achieve and as long as we can do that I'm fine with the conclusions.  

I understand the underlying frustrations and issues involved with locale and timezones with respect to data integration. It's the bane of a lot of data integration projects.  Some projects and environments will get data from all over the place with times in UTC, some will have a mix, some will be fixed for a certain country and timezone.  My feeling as such was that if there's a solution to be implemented, that it should be at the level of a project or project lifecycle environment.

That being said, I'm not against making it also an option in the Hop GUI and part of the new Hop installation configurator. Language, locale and timezone are perhaps important enough to ask about and set as a default for the developer.  We can take a big step forward in clarifying the situation.

**Comment by sramazzina** [comment]
I will consider your examples @hansva in our tests. Our implemention will solve this weird issues of headless cases but, if you want, you can preserve the things how they are now for compatibility by going through a configuration flag. I will push the PR in draft in the next hours so that everyone can take a look at how things are moving forward and comment on it so that we can come to a solution agreed by everybody. Thanks for your comments guys always much appreciated!
